backend/src/middleware/validation/validationMiddleware.js

```diff
@@ -57,11 +57,40 @@ const validateUpdatePost = async (resolve, root, args, context, info) => {
   return validatePost(resolve, root, args, context, info)
 }
 
+const validateReport = async (resolve, root, args, context, info) => {
+  const { resourceId } = args
+  const { user, driver } = context
+  if (resourceId === user.id) throw new Error('You cannot report yourself!')
+  const session = driver.session()
+  const reportQueryRes = await session.run(
+    `
+      MATCH (:User {id:$submitterId})-[:REPORTED]->(resource {id:$resourceId}) 
+      RETURN labels(resource)[0] as label
+    `,
+    {
+      resourceId,
+      submitterId: user.id,
+    },
+  )
+  const [existingReportedResource] = reportQueryRes.records.map(record => {
+    return {
+      label: record.get('label'),
+    }
+  })
+
+  if (existingReportedResource)
+    throw new Error(
+      `You have already reported the ${existingReportedResource.label}, please only report the same ${existingReportedResource.label} once`,
+    )
+  return resolve(root, args, context, info)
+}
+
 export default {
   Mutation: {
     CreateComment: validateCommentCreation,
     UpdateComment: validateUpdateComment,
     CreatePost: validatePost,
     UpdatePost: validateUpdatePost,
+    report: validateReport,
```

Comment:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-10-14T19:31:30Z" title="Monday, October 14th 2019, 9:31:30 pm +02:00">Oct 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -57,11 +57,40 @@ const validateUpdatePost = async (resolve, root, args, context, info) => {
   return validatePost(resolve, root, args, context, info)
 }
 
+const validateReport = async (resolve, root, args, context, info) => {
+  const { resourceId } = args
+  const { user, driver } = context
+  if (resourceId === user.id) throw new Error('You cannot report yourself!')
+  const session = driver.session()
+  const reportQueryRes = await session.run(
+    `
+      MATCH (:User {id:$submitterId})-[:REPORTED]->(resource {id:$resourceId}) 
+      RETURN labels(resource)[0] as label
+    `,
+    {
+      resourceId,
+      submitterId: user.id,
+    },
+  )
+  const [existingReportedResource] = reportQueryRes.records.map(record => {
+    return {
+      label: record.get('label'),
+    }
+  })
+
+  if (existingReportedResource)
+    throw new Error(
+      `You have already reported the ${existingReportedResource.label}, please only report the same ${existingReportedResource.label} once`,
+    )
+  return resolve(root, args, context, info)
+}
+
 export default {
   Mutation: {
     CreateComment: validateCommentCreation,
     UpdateComment: validateUpdateComment,
     CreatePost: validatePost,
     UpdatePost: validateUpdatePost,
+    report: validateReport,
```


Great done @mattwr18

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-14T19:33:20Z" title="Monday, October 14th 2019, 9:33:20 pm +02:00">Oct 14, 2019</time>_

---


glad you like it, I need to go back and do this for some of my `PR`s hahaha